test/unit/math/prim/mat/fun/qr_thin_R_test.cpp

```diff
@@ -0,0 +1,36 @@
+#include <stan/math/prim/mat.hpp>
+#include <gtest/gtest.h>
+
+TEST(MathMatrix, qr_thin_R) {
+  stan::math::matrix_d m0(0, 0);
+  stan::math::matrix_d m1(4, 2);
+  m1 << 1, 2, 3, 4, 5, 6, 7, 8;
+
+  using stan::math::qr_thin_Q;
+  using stan::math::qr_thin_R;
+  using stan::math::transpose;
+  EXPECT_THROW(qr_thin_R(m0), std::invalid_argument);
+  EXPECT_NO_THROW(qr_thin_R(m1));
+
+  stan::math::matrix_d m2(4, 2);
+  stan::math::matrix_d Q(4, 2);
+  stan::math::matrix_d R(2, 2);
+  Q = qr_thin_Q(m1);
+  R = qr_thin_R(m1);
+  m2 = Q * R;
+  for (unsigned int i = 0; i < m1.rows(); i++) {
+    for (unsigned int j = 0; j < m1.cols(); j++) {
+      EXPECT_NEAR(m1(i, j), m2(i, j), 1e-12);
+    }
+  }
+  for (unsigned int j = 0; j < m1.cols(); j++)
+    EXPECT_TRUE(R(j, j) >= 0.0);
+
+  stan::math::matrix_d m3(2, 4);
```

Comment:
I guess the only test that's missing is for a matrix where ncols == nrows. Is m3 filled with values more interesting than zeroes here?

Reply:
I added a test with a square matrix that has values `-(1:9)`, which is singular. That is fine because the QR decomposition does not assume anything about the input matrix.